### ISSUE TITLE
Bug: Cors Policy Execution Failed

### ISSUE THREAD

**Issue by thabaum** [oqtane/oqtane.framework#3675]
## Description
Cors policy is not executed when Oqtane.Server project starts up.
Microsoft.AspNetCore.Cors.Infrastructure.CorsService: Information: CORS policy execution failed.
![image](https://github.com/oqtane/oqtane.framework/assets/13577556/6c856813-c6dd-42d5-9a5b-4510307da789)

No Cors policy set on the file controller.  Access to the file controller should only work with a Cors Policy to enable it.

## Environment
Visual Studio 2022

## Solution
I may be wrong, but potentially changing the property type used for the policy name to a constant could resolve this issue.

I would also suggest investigating the middleware execution load order and the use of the `[EnableCors(Constants.MauiUserAgent)]` on the file controller itself.

## Test
Be sure to test file upload in .NET Maui to review results.

You should have to put `[EnableCors(Constants.MauiUserAgent)]` on the `[HttpPost("upload")]` file controller API method in order to access it, or it will(should) throw an exception error relating to not having permissions for this action to be allowed.
## Additional Information
Cors current middleware should generally execute after app.UseRouting() for this to work properly as well from testing I have done in the past.  

Instead of the current code:
```
            app.UseHttpsRedirection();
            app.UseStaticFiles();
            app.UseCors(Constants.MauiUserAgent);
            app.UseTenantResolution();
            app.UseJwtAuthorization();
            app.UseBlazorFrameworkFiles();
            app.UseRouting();
            app.UseAuthentication();
            app.UseAuthorization();
```
With all the configuration logic above I believe should look more like this:

```
            app.UseHttpsRedirection();
            app.UseStaticFiles();
            app.UseTenantResolution();
            app.UseJwtAuthorization();
            app.UseBlazorFrameworkFiles();
            app.UseRouting();
            app.UseCors(Constants.MauiUserAgent);  // move here
            app.UseAuthentication();
            app.UseAuthorization();
            app.UseAntiforgery();  // add this
```

at line 187 in startup.cs.

Really since it is not executing no way to really test if it is doing anything. It is not functioning stopping anything.  You can probably remove it with same file upload results in .NET Maui.

Hopefully this helps, maybe it is my IDE.  The file upload does work now with the Jwt Token approach, but I think there maybe a better solution here to test if it resolves upload issues alone and make sure the security features relating to Antiforgery and CORS in .NET in Oqtane perform as expected.  I will make another issue relating to Antiforgery updates once I can find ways to strengthen that case.

> app.UseAntiForgery() is not needed as the use of the Jwt token results in the anti-forgery validation being bypassed (anti-forgery is only needed for cookies - not tokens)

You would not need Jwt Tokens if this worked properly I do not believe.

I am caught in a pickle between Jwt Tokens and Cookies, but I believe that actually cookies is the focus moving forward for .NET Blazor?  Tokens for third party apps connecting to API, and cookies for everything else maybe?  Sorry to go a little off topic here I am paying close attention when these get discussed.  I know this project works hard to align with Microsoft's .NET Core Blazor project the best it can.  At the same time there is some very complex multi-tenant dynamic modular logic happening to hold the Oqtane Framework together.

**Comment by sbwalker** [comment]
@thabaum the reported issue is that a "Cors policy execution failed" message is logged in the server console during startup. I was able to reproduce this behavior:

```
Microsoft.AspNetCore.Hosting.Diagnostics: Information: Request starting HTTP/1.1 POST http://localhost:44357/_blazor/negotiate?negotiateVersion=1 - - 0
Microsoft.AspNetCore.Cors.Infrastructure.CorsService: Information: CORS policy execution failed.
Microsoft.AspNetCore.Cors.Infrastructure.CorsService: Information: Request origin http://localhost:44357 does not have permission to access the resource.


Microsoft.AspNetCore.Hosting.Diagnostics: Information: Request starting HTTP/1.1 GET http://localhost:44357/_blazor?id=1Iql700BiCTE3v03eaK0HA - - -
Microsoft.AspNetCore.Cors.Infrastructure.CorsService: Information: CORS policy execution failed.
Microsoft.AspNetCore.Cors.Infrastructure.CorsService: Information: Request origin http://localhost:44357 does not have permission to access the resource.
```

Note that this has nothing to do with .NET MAUI as this is happening when the framework starts up and the only client is the local browser. The errors seem to be related to requests to the "_blazor" path - which is part of Blazor.

I tested the changes in PR #3674 and they have no effect - the above "Cors policy execution failed" messages are still logged in the console. Further research indicates that the "Cors policy execution failed" message is a known problem and does not affect the run-time characteristics of a Blazor application. So it can be safely ignored.

So basically the "Cors policy execution failed" message is not a problem, so I will close this issue.

In regards to PR #3674 I conducted some further testing...

The current location for the UseCors() method in Startup.cs works fine. I tested the change in #3674 to move the location and it continues to work fine. I have verified that the Microsoft guidance does indicate that the UseCors() call should be after UseRouting() and before UseAuthorization() - so I do believe this change is valid (even if it has no effect).

This being said, if the goal is to only enable CORS for the specific API action methods which require it (ie. least privilege) then the UseCors(policyName) method should not be used in Startup - as it means the named CORS policy is applied GLOBALLY to the entire application. Instead the generic UseCors() method should be used which only registers the middleware but does not apply any specific policy. When using this approach you must use the [EnableCors] attribute on specific methods and specify the policyName. 

You are correct that adding [EnableCors] to the UploadFile method requires a constant for the policy name. It is probably best to create a dedicated constant for this purpose.

The addition of UseAntiForgery() is not needed. The framework already adds an anti-forgery token through _Host.cshtml so that every browser-based request is secured with an antiforgery token. It is added using the GetAndStoreTokens() method so that it is able to retrieve the ant-forgery token value and pass it to the Client application so that it can be used in client originated requests such as FORM POST actions through JavaScript (which login relies on). 

Note that anti-forgery is ONLY used in conjunction with cookies. In the case of an external client application (such as .NET MAUI) it is not possible to use cookies because there is no standard browser/server interaction. In these scenarios you need to use Jwt tokens. With Jwt tokens there is no need for anti-forgery validation and therefore it is bypassed. 

Lastly, there is no simple way in Blazor Hybrid to pass cookies from the client to the server - as the WebView does not behave like a standard browser. So if you need to pass a user authentication context you need to use a Jwt token. In the case of a file upload, authentication is a critical requirement - so this is why a Jwt token must be used. Enabling antiforgery has no benefit in this scenario as there are no cookies for antiforgery to validate.

The changes above are included in #3677

**Comment by thabaum** [comment]
@sbwalker thank you for taking time to review my research on the topic to better understand.  To note somehow I was able to make the file uploads work without the Jwt approach, not that the Jwt method is not going to work or isn't the best solution.

Now that these changes have merged after the doc meeting in a few I will adjust the source with the new and old JS interop and test along with checkup on all other areas now that we have Cors functioning as intended.  I am guessing that if I don't set a secret, I will be pretty much using the old method for file uploads since the secret does not exist.

Reason is I made an adjustment to the Antiforgery configuration I believe to make it work, I wish I could locate that code change easily as it worked once then quit (site settings bug).  I did have to add the add.UseAntiforgery() to the framework, and then I added some other configurations.  Now you have cleared up where exactly the antiforgery token is applied to the framework I can research it's behavior more so I can fully understand how it works.

It seemed like until I made a change I had no real control over antiforgery tokens behavior, then after making a few changes I did and could upload.  I may have created a bypass for that one file controller method based on the controller address so maybe not the best solution it is hard to recall.  I dropped that approach due to that site settings bug fooling me more or less.  It was my final test, then I went to help support the site settings bug since it related.

I didn't even know that User Settings Jwt token area even existed in Oqtane as I have been so focused on specifics.  That is a super powerful feature I look forward to working with.  And now I know more on how the antiforgery tokens for cookies work in Oqtane.

Thank you for taking the extra time to smooth out this Cors issue I greatly appreciate it.  When I create my own controllers I really want to understand 100% all security related features, the does and the don'ts while working with the Oqtane Framework.

If I find anything regarding Antiforgery to work on I will generate an issue specific for it with.

Cheers!
